corehq/apps/data_interfaces/tests/test_models.py

```diff
@@ -0,0 +1,253 @@
+        })
+
+    def test_to_dict_fills_in_missing_fields(self):
+        action = CaseRuleAction()
+
+        self.assertEqual(action.to_dict(), {
+            'update_case_definition': None,
+            'custom_action_definition': None,
+            'create_schedule_instance_definition': None,
+            'case_deduplication_action_definition': None
+        })
+
+
+class AutomaticUpdateRuleTests(SimpleTestCase):
+    def test_to_json_includes_all_fields(self):
+        rule = AutomaticUpdateRule(
+            domain='test-domain',
+            name='test-name',
+            case_type='test-case',
+            active=True,
+            deleted=False,
+            last_run=date(2020, 2, 20),
+            filter_on_server_modified=True,
+            server_modified_boundary=5,
+            workflow='test-workflow',
+            locked_for_editing=False,
+            upstream_id='upstream_id',
+            id=15,
+        )
+
+        self.assertEqual(rule.to_json(), {
+            'domain': 'test-domain',
+            'name': 'test-name',
+            'case_type': 'test-case',
+            'active': True,
+            'deleted': False,
+            'last_run': date(2020, 2, 20),
+            'filter_on_server_modified': True,
+            'server_modified_boundary': 5,
+            'workflow': 'test-workflow',
+            'locked_for_editing': False,
+            'upstream_id': 'upstream_id',
+            'id': 15,
+        })
+
+    def test_to_dict_includes_criteria_and_actions(self):
+        rule = AutomaticUpdateRule()
+        patch.object(rule, 'to_json', lambda: 'rule_data').start()
+
+        self.criteria.extend([
+            create_dict_mock(CaseRuleCriteria, 'criteria1'),
+            create_dict_mock(CaseRuleCriteria, 'criteria2'),
+        ])
+
+        self.actions.extend([
+            create_dict_mock(CaseRuleAction, 'action1'),
+            create_dict_mock(CaseRuleAction, 'action2'),
+        ])
+
+        self.assertEqual(rule.to_dict(), {
+            'rule': 'rule_data',
+            'criteria': ['criteria1', 'criteria2'],
+            'actions': ['action1', 'action2'],
+        })
+
+    def setUp(self):
+        self.actions = []
+        self.criteria = []
+
+        criteria_set_mock = MagicMock()
+        criteria_set_mock.all = lambda: self.criteria
+        criteria_patcher = patch.object(AutomaticUpdateRule, 'caserulecriteria_set', criteria_set_mock)
+        criteria_patcher.start()
+        self.addCleanup(criteria_patcher.stop)
+
+        action_set_mock = MagicMock()
+        action_set_mock.all = lambda: self.actions
+        action_patcher = patch.object(AutomaticUpdateRule, 'caseruleaction_set', action_set_mock)
+        action_patcher.start()
+        self.addCleanup(action_patcher.stop)
+
+
+def create_dict_mock(type, data):
+    '''Create an empty object of the given type whose `to_dict` method returns the specified data'''
+    obj = type()
```

Comment:
Nit response: `klass` is more conventional than `class_`, right?

Reply:
PEP-8 says the convention is to use a single trailing underscore

> `single_trailing_underscore_`: used by convention to avoid conflicts with Python keyword, e.g.
>  ```
> tkinter.Toplevel(master, class_='ClassName')
> ```

https://peps.python.org/pep-0008/#descriptive-naming-styles